### ISSUE TITLE
CS8321/IDE0051 false negative on recursive functions

### ISSUE THREAD

**Issue by NextTurn** [dotnet/roslyn#39406]
**Version Used**: Visual Studio 2019 Version 16.4

**Steps to Reproduce**:

```cs
class Program
{
    public static void Main()
    {
        int i;
        void Local(int _) => Local(i);

        static void StaticLocal() => StaticLocal();
    }

    private void Method() => this.Method();

    private static void StaticMethod() => StaticMethod();
}
```

**Expected Behavior**:

CS8321 is reported on `Local` and `StaticLocal`.
IDE0051 is reported on `Method` and `StaticMethod`.

**Actual Behavior**:

CS8321 is not reported on `Local` and `StaticLocal`.
IDE0051 is not reported on `Method` and `StaticMethod`.

**Comment by mavasani** [comment]
This does not look to be a bug. The symbols are indeed use, and hence the dead code diagnostics do not fire. You are probably looking for a different analyzer https://github.com/dotnet/roslyn-analyzers/issues/1478

**Comment by Youssef1313** [comment]
@mavasani Infinite recursion looks like a different case. For example:

```csharp
using System;

public class C
{
    private static int FactorialMethodThatIsUnused(int x)
    {
        if (x < 0) throw new ArgumentOutOfRangeException(nameof(x));
        if (x == 0) return 0;
        return x * FactorialMethodThatIsUnused(x - 1);
    }
}
```

While the method `FactorialMethodThatIsUnused` is technically used, but it's only used inside of itself. The analyzer should ideally report a diagnostic for it (unless the implementation would be much complicated).
